src/test/java/org/analytics/hydrator/plugin/TopNTest.java

```diff
@@ -133,62 +117,47 @@ private void testSingleEngine(Engine engine, String topField, String topSize, St
-      Assert.fail("Actual result: " + recordListToString(actual)
-                    + " does not match expected: " + recordListToString(expected));
-    }
+    // Records from sink are out of order, so use Sets to compare
+    Assert.assertEquals(Sets.newHashSet(expected), Sets.newHashSet(actual));
   }
 
-  private String recordListToString(List<StructuredRecord> recordList) {
-    List<String> recordStringList = Lists.transform(recordList, new Function<StructuredRecord, String>() {
-      @Override
-      public String apply(final StructuredRecord record) {
-        return recordToString(record);
-      }
-    });
-    return "{" + COMMA_JOINER.join(recordStringList) + "}";
-  }
+  @Test
+  public void testAllNumericFields() throws Exception {
+    // Sort records with field "age" of int type and skip records with null value in field "age"
+    testTopN("age", 4, true, "skipNull", ImmutableList.of(ALICE, EVE, LEO));
 
-  private String recordToString(final StructuredRecord record) {
-    List<String> fieldValues= Lists.transform(SCHEMA.getFields(), new Function<Schema.Field, String>() {
-      @Override
-      public String apply(Schema.Field field) {
-        return field.getName() + "=" + record.get(field.getName());
-      }
-    });
-    return "[" + COMMA_JOINER.join(fieldValues) + "]";
-  }
+    // Sort records with field "age" of int type and keep records with null value in field "age"
+    testTopN("age", 4, false, "keepNull", ImmutableList.of(ALICE, EVE, LEO, BOB_NULL_AGE));
 
-  @Test
-  public void test() throws Exception {
-    // Sort records with field "age" of int type and skip null
-    testTopN("age", "4", "true", "skipNull", ImmutableList.of(ALICE, EVE, LEO));
+    // Sort records with field "id" of long type
+    testTopN("id", 2, false, "largest", ImmutableList.of(ALICE, BOB_NULL_AGE));
 
-    // Sort records with field "age" of int type and keep null
-    testTopN("age", "4", "false", "keepNull", ImmutableList.of(ALICE, EVE, LEO, BOB_NULL_AGE));
+    // Sort records with field "kg" of double type
+    testTopN("kg", 2, false, "heaviest", ImmutableList.of(ALICE, BOB_NULL_AGE));
 
-    // topField with non-numeric type is invalid
-    try {
-      testTopN("name", "4", "false", "nonNumeric", new ArrayList<StructuredRecord>());
-      Assert.fail("topField with non-numeric type is not allowed and should throw IllegalStateException");
-    } catch (IllegalStateException e) {
-      // Expected to catch IllegalArgumentException because "name" field has String type
-    }
+    // Sort records with field "cm" of float type
+    testTopN("cm", 2, false, "tallest", ImmutableList.of(ALICE, BOB_NULL_AGE));
+  }
 
+  @Test
+  public void testFailOnNonExistField() throws Exception {
     // Non-existing topField should throw Eception
     try {
-      testTopN("nonExist", "4", "false", "nonExist", new ArrayList<StructuredRecord>());
+      testTopN("nonExist", 4, false, "nonExist", new ArrayList<StructuredRecord>());
```

Comment:
shouldn't be deploying an app to test validation logic. Move the validate method from the plugin to the config, and add a config test that just calls validate.